Include soft deleted in Deploy Reports

rikeoka opened the pull request:
* Include soft-deleted projects in Deploy CSV Report filters
* Include soft-deleted items in Deploy CSV Report
* Bug fixes in Jira
* Add real-time reports for deploys#search limited to 1000 records for RT download of what you see
* Add stage name and deploy groups names to the report
* Test that soft deleted project deploys are included in report and in the dropdown

/cc @zendesk/samson

### Tasks
 - [X] Add capability to report on soft_deleted projects and select soft_deleted projects
 - [x] Update filter to support deploy_groups.production
 - [x] Update filter to change ending date filter to be 23:59:59 instead of 00:00:00
 - [x] Work out job failures caused by soft_deleted environment.  
 - [x] Add in deploy groups & stage name to report
 - [x] Fast response deploys#search.csv reports limited to a maximum of 1000 deploys, enough for quick analysis reviews.  If over 1000, it indicates that you should go to the job executed report.
 - [X] Change "no code deployed" to "code deployed" in report
 - [ ] :+1: from team

### References
 - Jira link: https://zendesk.atlassian.net/browse/SAMSON-263

### Risks
- Level: Medium
- Modification to default scope for Deploy, Environment, and Stage.  Global variable $INCLUDE_DELETED removes the soft deletion scope when csv report is being generated in CsvExportJob.  It is unset upon completion and failures.

rikeoka commented:
@grosser / @jonmoter 

Updated the report to include soft deleted deploys in the report.

If there is a better way to do this let me know.  I tried other methods but it kept on applying the scope for the associations.

grosser commented on app/models/deploy.rb:
```diff
@@ -130,11 +130,31 @@ def self.expired
+  end
+
+  def r_buddy
+    User.unscoped { buddy_id > 0 ? User.find(buddy_id) : nil }
```
uncachabel ... how about `buddy || (buddy_id && User.with_deleted { User.find(buddy_id) })`

grosser replied:
also r_buddy is a strange name ... buddy_with_deleted ?

grosser commented on app/models/deploy.rb:
```diff
@@ -130,11 +130,31 @@ def self.expired
+  end
+
+  def r_summary
+    "#{summary_action} #{short_reference} to #{r_stage.name}"
```
add a option to summary to use either stage or stage_with_deleted

grosser commented on app/models/deploy.rb:
```diff
@@ -130,11 +130,31 @@ def self.expired
-    user.id == buddy_id ? "bypassed" : buddy.try(:email)
+    user.id == buddy_id ? "bypassed" : r_buddy.try(:email)
+  end
+
```
also all private ... and tests ...

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -30,7 +30,7 @@
       <div class="form-group">
         <%= label_tag :project, "Project", class: "col-lg-2 control-label" %>
-        <% projects = [['', 0]] + Project.pluck(:name, :id) %>
+        <% projects = [['', 0]] + Project.unscoped.pluck(:name, :id) %>
```
mark deleted projects with (deleted) and maybe sort them to the back

rikeoka commented on test/models/deploy_test.rb:
```diff
@@ -241,6 +255,32 @@ def deploy!
+    before { Stage.any_instance.stubs(:deploy_requires_approval?).returns true }
+
+    it "returns array with deleted object values" do
+      project.update_attributes(repository_url: "https://github.com/samson-test-org/example-project")
```
Had to do this to get it to pass the test.  This probably means that because of the validations in projects it will fail if the project is soft deleted and the repo is deleted.

Not sure if its possible to preload all the associated data into the object without associations where it skips any validations for reporting.

grosser replied:
`update_column(:deleted_at, Time.now)` ?

rikeoka replied:
The project fixture had an invalid repo url and was giving an invalid repo url error whenever the project was called.  Not sure if this will error in a real instance when the project and repo is deleted.  I can't reproduce it in a live account, probably because GitRepository already exists for the project.

rikeoka replied:
Ah got it now, it was doing the validation on the project, when I was soft_deleting it.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,22 +34,19 @@ def deploy_csv_export(csv_export)
 
-    @deploys = Deploy.joins(:stage, :job).where(filter)
+    @deploys = Deploy.joins('INNER JOIN stages ON stages.id = deploys.stage_id INNER JOIN jobs ON jobs.id =' \
+                            ' deploys.job_id').unscope(where: :deleted_at).where(filter)
```
2-space indent ?

grosser replied:
or
```
.joins(
   'long string
)
```

grosser commented on app/models/deploy.rb:
```diff
@@ -130,19 +135,41 @@ def self.expired
 
+  def to_csv
+    [id, project_with_deleted.name, summary(with_deleted: true), commit, job.status, updated_at,
+     start_time, user_with_deleted.try(:name), user_with_deleted.try(:email), buddy_name, buddy_email,
```
2-space indent plz

grosser commented on app/models/deploy.rb:
```diff
@@ -23,8 +23,13 @@ def cache_key
-    "#{job.user.name} #{deploy_buddy} #{summary_action} #{short_reference} to #{stage.name}"
+  def summary(with_deleted: false)
+    if with_deleted
+      "#{job.user.name} #{deploy_buddy(with_deleted: true)} #{summary_action} #{short_reference} to"\
```
always call `deploy_buddy(with_deleted: with_deleted)`

grosser commented on app/models/deploy.rb:
```diff
@@ -179,8 +206,12 @@ def validate_stage_uses_deploy_groups_properly
-    return unless stage.deploy_requires_approval?
+  def deploy_buddy(with_deleted: false)
+    if with_deleted
+      return unless stage_with_deleted.deploy_requires_approval?
```
always call `stage_with_deleted(with_deleted: with_deleted)`

rikeoka replied:
removed with_deleted: with_deleted option since to have valid information it should pull deleted stages.

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -30,7 +30,8 @@
         <%= label_tag :project, "Project", class: "col-lg-2 control-label" %>
-        <% projects = [['', 0]] + Project.pluck(:name, :id) %>
+        <% projects = [['', 0]] + Project.pluck(:name, :id) +
+           Project.unscoped.where("deleted_at IS NOT NULL").map { |p| ["(deleted) #{p.name}", p.id] } %>
```
still pluck and then map so it pulls down less data

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,22 +34,20 @@ def deploy_csv_export(csv_export)
-    @deploys = Deploy.joins(:stage, :job).where(filter)
+    @deploys = Deploy.joins(
+      'INNER JOIN stages ON stages.id = deploys.stage_id INNER JOIN jobs ON jobs.id = deploys.job_id'
+    ).unscope(where: :deleted_at).where(filter)
```
this select all deleted deploys now ... and not both deleted and not deleted ?

rikeoka replied:
This works the same as unscoped, but instead of removing the entire default_scope it just removes from the query "WHERE `deleted_at` IS NULL" condition.  This was when I was trying to join with the associations using joins(:jobs, :stages) but fails since this only applies on the main Model and not associations.

rikeoka replied:
With the different association, optimized the query to prefetch associations during the join except for users and buddy.  Otherwise the in-memory table probably gets a bit large.

grosser commented on app/models/deploy.rb:
```diff
@@ -130,19 +130,43 @@ def self.expired
   private
 
+  def buddy_with_deleted
+    buddy || (buddy_id && User.with_deleted { User.find(buddy_id) })
```
need to cache all of these so they do not cause multiple queries ... also some tests for 'does no query when there is no id' ... 

```
return @foo if defined?(@foo)
@foo = bla || with_delete.find_by_id(xxx)
```

most likely also find_by_id so we don't blow up if something was hard-deleted ...

grosser commented on app/models/deploy.rb:
```diff
@@ -130,19 +130,43 @@ def self.expired
+    Project.unscoped { Project.find(stage_with_deleted.project_id) }
+  end
+
+  def stage_with_deleted
```
should do the same `stage ||` dance to leverage the pre-loaded stages -> avoid n+1 queries

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -30,7 +30,8 @@
         <%= label_tag :project, "Project", class: "col-lg-2 control-label" %>
-        <% projects = [['', 0]] + Project.pluck(:name, :id) %>
+        <% projects = [['', 0]] + Project.pluck(:name, :id) +
+           Project.unscoped.where("deleted_at IS NOT NULL").pluck(:name, :id).map { |p| ["(deleted) #{p[0]}", p[1]] } %>
```
`unscoped.where("deleted_at IS NOT NULL")` -> `scope :deleted`

grosser commented on app/models/deploy.rb:
```diff
@@ -130,19 +130,43 @@ def self.expired
+  def user_with_deleted
+    User.unscoped { job.user_id > 0 ? User.find(job.user_id) : nil }
+  end
+
```
instead of repeating this logic maybe do a concern like:
```Ruby
def with_deleted(association)
  @with_deleted ||= {}
  @with_deleted.fetch(object) do
    found = 
    @with_deleted[object] = send(object) || (id = send("#{object}_id") && association(object).klass.unscoped.find_by_id(id) }
end
```

grosser replied:
I guess for now the copy-paste is good enough ... but at least some tests for the edge cases

rikeoka replied:
would it make sense to do an association instead?
belongs_to :object_with_deleted, -> { unscope(where: 'deleted_at') }, -> class: 'Object', foreign_key: 'object_id

At least it probably makes sense in certain cases like buddy to remove complexity

grosser replied:
that could work ... and would cache nicely ...

rikeoka replied:
Testing each of these associations is needed?  I think it's a bit overkill to do that.

grosser replied:
test on in-depth and the rest can be copy-paste ?

On Thu, Jul 14, 2016 at 9:47 AM, Robert Ikeoka <notifications@github.com>
wrote:

> In app/models/deploy.rb
> <https://github.com/zendesk/samson/pull/1131#discussion_r70840556>:
>
> > +  def buddy_with_deleted
> > +    buddy || (buddy_id && User.with_deleted { User.find(buddy_id) })
> > +  end
> > +
> > +  def project_with_deleted
> > +    Project.unscoped { Project.find(stage_with_deleted.project_id) }
> > +  end
> > +
> > +  def stage_with_deleted
> > +    Stage.unscoped { Stage.find(stage_id) }
> > +  end
> > +
> > +  def user_with_deleted
> > +    User.unscoped { job.user_id > 0 ? User.find(job.user_id) : nil }
> > +  end
> > +
>
> Testing each of these associations is needed? I think it's a bit overkill
> to do that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1131/files/be649820fdf2ba2f677a17a345a92967c5b8f28d#r70840556>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAsZ4q-fdKkOugNptHfcAHTlGntBkK9ks5qVmgpgaJpZM4JIewY>
> .
>

rikeoka commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,9 +34,8 @@ def deploy_csv_export(csv_export)
-    @deploys = Deploy.joins(
-      'INNER JOIN stages ON stages.id = deploys.stage_id INNER JOIN jobs ON jobs.id = deploys.job_id'
-    ).unscope(where: :deleted_at).where(filter)
+    @deploys = Deploy.includes(:job, {stage_with_deleted: :project}).joins(:job, { stage_with_deleted: :project })
```
Should this be done with select to grab only what's needed from the db to reduce amount of data in memory?

grosser replied:
I'd imagine it only instantiates each project once, so this might be not that bad

rikeoka commented:
@grosser fixed all the comments.  Let me know if its ready for a squash.

grosser commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -71,9 +71,18 @@ def deploy_filter
-      when 'No'  then filter['stages.production'] = false
+      if production == 'Yes'
+        if DeployGroup.enabled?
+          filter['environments.production'] = true
```
logic duplication, can dry up with:
`filter_key = (DeployGroup.enabled? ? 'environments.production' : 'stages.production')`

and then 

```
filter[filter_key] = true
```

grosser replied:
some docs like '# we join different tables when using deploy groups' might be good too

grosser commented on app/controllers/deploys_controller.rb:
```diff
@@ -178,4 +184,19 @@ def find_deploy
+
+  # Creates a CSV for @deploys as a result of the search query limited to 1000 for speed
+  def realtime_csv
+    csv_limit = (params[:limit].presence || 1000).to_i
```
users can send in 99999 ->
````
max = 1000
csv_limit = [(params[:limit].presence || max).to_i, max].min
```

grosser commented on app/controllers/deploys_controller.rb:
```diff
@@ -178,4 +184,19 @@ def find_deploy
+    csv_limit = (params[:limit].presence || 1000).to_i
+    CSV.generate do |csv|
+      csv << Deploy.csv_header
+      @deploys.limit(csv_limit).each { |deploy| csv << deploy.csv_line }
```
.as_csv could be a nice method name ...

grosser commented on app/controllers/deploys_controller.rb:
```diff
@@ -178,4 +184,19 @@ def find_deploy
+      @deploys.limit(csv_limit).each { |deploy| csv << deploy.csv_line }
+      csv << ['-', 'count:', @deploys.limit(csv_limit).count]
+      csv << ['-', 'params:', params]
+      if @deploys.count > csv_limit
```
could save that .count query (which will be expensive with lots of deploys) and ask for 1001 and then check if we got 1001 ... and always pop 1

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,28 +34,44 @@ def deploy_csv_export(csv_export)
     filter = csv_export.filters
 
-    @deploys = Deploy.joins(:stage, :job).where(filter)
+    get_deploys(filter)
```
avoid get_ / set_ methods ... we are not in java land :)

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,28 +34,44 @@ def deploy_csv_export(csv_export)
   end
 
+  def get_deploys(filter)
+    if filter.keys.include?('environments.production')
```
can do a `@deploys = filter_deploys` so this method does not need to know about `@deploys` and dry it up a bit

grosser replied:
so this is the `searching for deleted` test ? ... can we make that more obvious ? :)

rikeoka replied:
Adding another grouping around the tests to make it clear with DeployGroups and without.  Wasn't the searching for deleted test.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,28 +34,44 @@ def deploy_csv_export(csv_export)
 
+  def get_deploys(filter)
+    if filter.keys.include?('environments.production')
+      production_value = filter.delete('environments.production')
```
csv_exports_controller sets 2 different keys and this only supports 1 key ... potential bug ?

grosser replied:
ah nvm checking that above

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,28 +34,44 @@ def deploy_csv_export(csv_export)
+      # To match logic of stages.production? True when any deploy_group environment is true or
+      # deploy_groups environment is empty and stages is true
+      production_query = "(A.production = ? OR (A.production IS NULL AND stages.production = ?))"
+      production_query = "NOT " + production_query if production_value == false
```
FYI `production_query.prepend "NOT "`

grosser replied:
A for a table name is not very intuitive ... maybe deploys_left or something like that ?

grosser replied:
no wait ... they both query only for deleted records ???

rikeoka replied:
Moving the subquery out to a pre-constructed string and labeling it "StageProd" 

stage.production? returns true if any DeployGroup.Environment.production is true or if no Deploy Groups are associated with the stage, Stage.production.

production = true, The query is returning StageProd.production is true or if StageProd is not present the stages.production is true.

production = false, probably could use the same logic with the false values, but there may be some conditions where NOT could give the undefined values as well.

stage_prod null = no associated environment.production for the stage (can never be null from db level validation)
alt logic is not using not and passing false on instead of true

stage | stage_prod | expected from stage.production? | this logic | alt logic
----- | ------------ | --------------------------------- | --------- | --------
any | true | true | true | true
any | false | false | false | false
true | null | true | true | true
false | null | false | false | false
null | null | false | false | never applicable

rikeoka replied:
For some reason this didn't work out the way it should, so going to redo the logic on the filter to get expected.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,28 +34,44 @@ def deploy_csv_export(csv_export)
+          "INNER JOIN environments ON environments.id = deploy_groups.environment_id) A ON A.stage_id = stages.id").
+        unscope(where: :deleted_at).where(filter).where(production_query, true, true)
+    else
+      @deploys = Deploy.includes(:buddy, job: :user, stage_with_deleted: :project).
```
does that stage_with_deleted work ... I though you said it does not ....

rikeoka replied:
This works for preloading the associated objects.  What doesn't work is the use of it to automatically join the tables for the filter to work.  So i can't do Deploy.includes(:stages_with_deleted).joins(:stages_with_deleted) but I can do Deploy.includes(:stages_with_deleted).joins("INNER JOIN ...")

rikeoka replied:
With this I was getting the expected output report, but had a performance improvement in generation time 10s for 20,000 deploys vs 28 seconds without.

grosser commented on app/models/csv_export.rb:
```diff
@@ -31,11 +31,18 @@ def filters
   end
 
+  def filters_project
+    if filters['stages.project_id']
```
can avoid duplication with `if id = filters['stages.project_id']`

grosser commented on app/models/csv_export.rb:
```diff
@@ -31,11 +31,18 @@ def filters
+  def filters_project
+    if filters['stages.project_id']
+      proj = Project.unscoped.where(id: filters['stages.project_id']).first.try(:permalink)
+      proj + '_' unless proj.nil?
```
`unless proj.nil?` -> `if proj`

grosser commented on app/models/deploy.rb:
```diff
@@ -4,7 +4,8 @@ class Deploy < ActiveRecord::Base
   belongs_to :job
-  belongs_to :buddy, class_name: 'User'
+  belongs_to :buddy, -> { unscope(where: "deleted_at") }, class_name: 'User'
+  belongs_to :stage_with_deleted, -> { unscope(where: "deleted_at") }, class_name: 'Stage', foreign_key: 'stage_id'
```
isn't this `deleted_stage` ... since it only filters those that are deleted ?

rikeoka replied:
It removes the "WHERE stages.deleted_at IS NULL" from the query.  So it would get both non-deleted and deleted stages.  If the desired is deleted stage, it would need unscope(where: "deleted_at").where("stages.deleted_at IS NOT NULL") instead.  

I didn't want to update the regular stages scope, since that would cause some wierd UI bugs, but changing the buddy association was fine, since generally if people are looking back in time, they would still want to know who the buddy is even though the buddy is no longer an active user in Samson.

grosser replied:
uh nice ... so it removes the scope on that column ... funky

On Wed, Jul 27, 2016 at 7:16 PM, Robert Ikeoka <notifications@github.com>
wrote:

> In app/models/deploy.rb
> <https://github.com/zendesk/samson/pull/1131#discussion_r72556016>:
>
> > @@ -4,7 +4,8 @@ class Deploy < ActiveRecord::Base
> >    belongs_to :stage, touch: true
> >    belongs_to :build
> >    belongs_to :job
> > -  belongs_to :buddy, class_name: 'User'
> > +  belongs_to :buddy, -> { unscope(where: "deleted_at") }, class_name: 'User'
> > +  belongs_to :stage_with_deleted, -> { unscope(where: "deleted_at") }, class_name: 'Stage', foreign_key: 'stage_id'
>
> It removes the "WHERE stages.deleted_at IS NULL" from the query. So it
> would get both non-deleted and deleted stages. If the desired is deleted
> stage, it would need unscope(where: "deleted_at").where("stages.deleted_at
> IS NOT NULL") instead.
>
> I didn't want to update the regular stages scope, since that would cause
> some wierd UI bugs, but changing the buddy association was fine, since
> generally if people are looking back in time, they would still want to know
> who the buddy is even though the buddy is no longer an active user in
> Samson.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1131/files/188714baf38f6a4a9cadec2f58016756e4f9262d#r72556016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ61G4rL4prRBzzZM-ScJSFovnOjUks5qaBD7gaJpZM4JIewY>
> .
>

grosser commented on app/controllers/deploys_controller.rb:
```diff
@@ -178,4 +184,22 @@ def find_deploy
+    max = 1000
+    csv_limit = [(params[:limit].presence || max).to_i, max].min
+    # may be expensive to count an unfiltered @deploys so do count once
+    deploy_count = @deploys.limit(csv_limit + 1).count
```
don't think you can limit a count ... also still doing 2 queries ... 

deploys = @deploys.limit(x+1).to_a
deploys[0..-2].each do |deploy|
if deploys.length > limit

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,28 +34,50 @@ def deploy_csv_export(csv_export)
+  def filter_deploys(filter)
+    if filter.keys.include?('environments.production')
+      production_value = filter.delete('environments.production')
+      # To match logic of stages.production? True when any deploy_group environment is true or
```
best add a comment to `def production?` too ... `# this is reproduced via sql in xxx.rb, make them sync when editing`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -34,28 +34,50 @@ def deploy_csv_export(csv_export)
+      production_value = filter.delete('environments.production')
+      # To match logic of stages.production? True when any deploy_group environment is true or
+      # deploy_groups environment is empty and stages is true
+      production_query = "(StageProd.production = ? OR (StageProd.production IS NULL AND stages.production = ?))"
```
should be in a if/else block since it is just overwritten when `production_value == false`

grosser commented on app/models/job.rb:
```diff
@@ -1,6 +1,6 @@
 class Job < ActiveRecord::Base
   belongs_to :project
-  belongs_to :user
+  belongs_to :user, -> { unscope(where: 'deleted_at') }
```
does `unscoped(where: :deleted_at)` also work ?

rikeoka replied:
unscoped doesn't work.  It works as a class method to remove all scopes.  In associations unscoped doesn't seem to work at all, and only the query methods seems to work.

grosser commented:
hmm any chance we can do this whole with_deleted without adding new methods everywhere and do `Stage.with_deleted { ... generate csv ... }` instead ?

grosser commented on app/models/stage.rb:
```diff
@@ -174,6 +180,10 @@ def destroy
   end
 
+  def deploy_group_names
+    deploy_groups.pluck(:name).join('|')
```
this seems weird ... should return an array ... then users can join them as they want ...
also needs `.sort_by(&:natural_order)` ... could make `natural_order` static and call it or do a `deploy_groups.select(:name).sort_by(&:natural_order).map!(&:name)`

grosser commented on test/controllers/deploys_controller_test.rb:
```diff
@@ -168,6 +168,19 @@ def self.with_and_without_project(&block)
         assert_response :ok
       end
 
+      it "it renders csv" do
```
`it it`

grosser commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -211,21 +211,28 @@ def self.show_test(state)
+        end
       end
 
+      prod_filter_test("Yes", true)
```
could be nicer with `it_filters_production "Yes", true`

rikeoka commented:
@grosser updated to address the comments.

Unfortunately with the default scope set on the model, the associations will always apply the default scope on associations even with AssociatedClass.with_deleted or unscoped.
Ways around this would be Class.unscoped.find(class_id) but where data is being gathered from a method, it would likely require duplicating the logic of the method.  This probably is the cleanest way as it fails nicely when the data can't be found.  Ie can't do Project.with_deleded { stage.project.name } the project query will still apply the WHERE projects.delete_at IS NULL statement. 

For deploy.csv_line it would mean to get deploy.project_with_deleted.name attribute, it would need a statement like Project.unscoped.where(id: Stage.unscoped.where(id: stage_id).project_id).first.try(:name) unless Stage.unscoped.where(id: stage_id).first.project_id.nil?

I'm open to suggestions.  Maybe raw SQL query?  I don't know if postgree and sqllite supports the GROUP_CONCAT syntax MySQL does.  Support for multiple db engines makes doing a complex query difficult to work across all of the supported db types.  BTW, when I ran the following raw SQL extract from mySQL it can dump 20k records in about 3 seconds versus 16-8 seconds through the Rails way.

SELECT deploys.id, projects.name, jobs.status, jobs.commi, deploys.updated_at, deploys.created_at, deployer.name, deployer.email, buddy.name, buddy.email, stages.name, (stages.production OR StageProd.production) as production, NOT stages.no_code_deployed as code_deployed, projects.deleted_at, DEPLOY_GROUPS.deploy_group_names
FROM deploys
INNER JOIN jobs ON jobs.id = deploys.job_id 
INNER JOIN stages ON stages.id = deploys.stage_id
INNER JOIN projects on projects.id = stages.project_id
LEFT JOIN (SELECT DISTINCT deploy_groups_stages.stage_id, environments.production 
  FROM deploy_groups_stages
  INNER JOIN deploy_groups ON deploy_groups.id = deploy_groups_stages.deploy_group_id
  INNER JOIN environments ON environments.id = deploy_groups.environment_id
) StageProd ON StageProd.stage_id = stages.id
LEFT JOIN (SELECT DISTINCT deploy_groups_stages.stage_id, GROUP_CONCAT(DISTINCT deploy_groups.name ORDER BY deploy_groups.name ASC SEPARATOR '|') as deploy_group_names
  FROM deploy_groups_stages 
  INNER JOIN deploy_groups on deploy_groups.id = deploy_groups_stages.deploy_group_id
  GROUP BY deploy_groups_stages.stage_id
) DEPLOY_GROUPS ON DEPLOY_GROUPS.stage_id = stages.id
LEFT JOIN users as deployer ON deployer.id = jobs.user_id
LEFT JOIN users as buddy ON buddy.id = deploys.buddy_id

grosser commented:
man that's some ugly query :D

... hmm since the default scope is a block we might be able to completely
disable it via some global setter ...

default_scope { $GIMME_ALL ? where(nil) : where(deleted_at: nil) }

$GIMME_ALL = true
export
$GIMME_ALL = false


On Fri, Jul 29, 2016 at 4:25 PM, Robert Ikeoka <notifications@github.com>
wrote:

> @grosser <https://github.com/grosser> updated to address the comments.
>
> Unfortunately with the default scope set on the model, the associations
> will always apply the default scope on associations even with
> AssociatedClass.with_deleted or unscoped.
> Ways around this would be Class.unscoped.find(class_id) but where data is
> being gathered from a method, it would likely require duplicating the logic
> of the method. This probably is the cleanest way as it fails nicely when
> the data can't be found. Ie can't do Project.with_deleded {
> stage.project.name } the project query will still apply the WHERE
> projects.delete_at IS NULL statement.
>
> For deploy.csv_line it would mean to get deploy.project_with_deleted.name
> attribute, it would need a statement like Project.unscoped.where(id:
> Stage.unscoped.where(id: stage_id).project_id).first.try(:name) unless
> Stage.unscoped.where(id: stage_id).first.project_id.nil?
>
> I'm open to suggestions. Maybe raw SQL query? I don't know if postgree and
> sqllite supports the GROUP_CONCAT syntax MySQL does. Support for multiple
> db engines makes doing a complex query difficult to work across all of the
> supported db types. BTW, when I ran the following raw SQL extract from
> mySQL it can dump 20k records in about 3 seconds versus 16-8 seconds
> through the Rails way.
>
> SELECT deploys.id, projects.name, jobs.status, jobs.commi,
> deploys.updated_at, deploys.created_at, deployer.name, deployer.email,
> buddy.name, buddy.email, stages.name, (stages.production OR
> StageProd.production) as production, NOT stages.no_code_deployed as
> code_deployed, projects.deleted_at, DEPLOY_GROUPS.deploy_group_names
> FROM deploys
> INNER JOIN jobs ON jobs.id = deploys.job_id
> INNER JOIN stages ON stages.id = deploys.stage_id
> INNER JOIN projects on projects.id = stages.project_id
> LEFT JOIN (SELECT DISTINCT deploy_groups_stages.stage_id,
> environments.production
> FROM deploy_groups_stages
> INNER JOIN deploy_groups ON deploy_groups.id =
> deploy_groups_stages.deploy_group_id
> INNER JOIN environments ON environments.id = deploy_groups.environment_id
> ) StageProd ON StageProd.stage_id = stages.id
> LEFT JOIN (SELECT DISTINCT deploy_groups_stages.stage_id,
> GROUP_CONCAT(DISTINCT deploy_groups.name ORDER BY deploy_groups.name ASC
> SEPARATOR '|') as deploy_group_names
> FROM deploy_groups_stages
> INNER JOIN deploy_groups on deploy_groups.id =
> deploy_groups_stages.deploy_group_id
> GROUP BY deploy_groups_stages.stage_id
> ) DEPLOY_GROUPS ON DEPLOY_GROUPS.stage_id = stages.id
> LEFT JOIN users as deployer ON deployer.id = jobs.user_id
> LEFT JOIN users as buddy ON buddy.id = deploys.buddy_id
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1131#issuecomment-236318120>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ6YZQhZ0f8CJ58dm0TcUUfaG6Zooks5qaovvgaJpZM4JIewY>
> .
>

rikeoka commented:
@grosser 

This works, but it may be a bit dangerous and cause UI glitches while a report is being generated.  At a minimum deleted Deploy, Environment, Project and Stage would be affected.  There is an off chance for instance that when the report is running a deleted stage could appear and a deploy is created for one.

Because of tweaking the default scope going to bump the risk for this to medium as well since it could cause unknown effects while reports are being run.

The test to show how this works for 3 cases nil, false, true is in models/deploy_test #stage  Stage had the quirkiest effects, I think paper_trail may add the soft_deletion scope on the model, but fortunately the unscope(with: 'deleted_at') works.  I didn't bother updating deploys#buddy or jobs#user since I think it makes sense that association is always unscoped for UI purposes.

grosser commented:
has_soft_deletion unless self < SoftDeletion::Core
+
same in plugins/pipelines/app/decorators/stage_decorator.rb
+
default_scope -> { Thread.current[:stage_with_deleted] ? where(nil) :
where(deleted_at: nil) }

and it works :D

-> no funky with_deleted methods / able to use joins(:environments) etc

... I'll update soft_deletion to use this since it's more universal then
unscoped/with_deleted ...

"SELECT `stages`.* FROM `stages` INNER JOIN `projects` ON `projects`.`id` =
`stages`.`project_id` AND `projects`.`deleted_at` IS NULL WHERE
`stages`.`deleted_at` IS NULL  ORDER BY `stages`.`order` ASC"
>> Thread.current[:stage_with_deleted] = true
=> true
>> Stage.joins(:project).to_sql
=> "SELECT `stages`.* FROM `stages` INNER JOIN `projects` ON
`projects`.`id` = `stages`.`project_id`  ORDER BY `stages`.`order` ASC"


... whereas the current approach does not work ...

Stage.with_deleted { Project.with_deleted { Stage.joins(:project).to_sql } }
=> "SELECT `stages`.* FROM `stages` INNER JOIN `projects` ON
`projects`.`id` = `stages`.`project_id` AND `projects`.`deleted_at` IS NULL"

On Fri, Jul 29, 2016 at 7:40 PM, Robert Ikeoka <notifications@github.com>
wrote:

> @grosser <https://github.com/grosser>
>
> This works, but it may be a bit dangerous and cause UI glitches while a
> report is being generated. At a minimum deleted Deploy, Environment,
> Project and Stage would be affected. There is an off chance for instance
> that when the report is running a deleted stage could appear and a deploy
> is created for one.
>
> Because of tweaking the default scope going to bump the risk for this to
> medium as well since it could cause unknown effects while reports are being
> run.
>
> The test to show how this works for 3 cases nil, false, true is in
> models/deploy_test #stage Stage had the quirkiest effects, I think
> paper_trail may add the soft_deletion scope on the model, but fortunately
> the unscope(with: 'deleted_at') works. I didn't bother updating
> deploys#buddy or jobs#user since I think it makes sense that association is
> always unscoped for UI purposes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1131#issuecomment-236335381>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ4aeFlK1Wc20mL0fgHg_Dj25bRRxks5qarmFgaJpZM4JIewY>
> .
>

rikeoka commented:
Cool.  It also works with this so a where(nil) isn't even needed.
default_scope -> { where(deleted_at: nil) unless Thread.current[:stage_with_deleted] }

grosser commented:
bump soft_deletion to 0.8.0 and you can get rid of all the extra model helpers and just use `User.with_deleted { Stage.with_deleted { ...` around your code

rikeoka commented:
Updated.  Squashed the last few experimental commits to show the cleanup of the unscoping calls better.

I'll leave it up to you if we want to apply the same to the Job.user and Deploy.buddy association, since it does make sense from a UI perspective to show "John Doe" deployed XYZ is better than "Deleted User" deployed XYZ.  Rest of them should be cleaned up.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -30,32 +30,66 @@ def generate_csv(csv_export)
 
+  # Relper method to removes the default soft_deletion scope for these models for the report
+  def remove_deleted_scope_and_create_report(csv_export)
+    Deploy.with_deleted do
```
this looks fun :D

grosser commented:
👍 so much better :D 
crazy sql is still a bit scary .... but at least it is contained ...

That buddy deletion might come back to bite us when someone joins on buddy and then something does not work as expected, but looks like a sane change ...
